src/main/java/appeng/api/networking/crafting/ICraftingCPU.java

```diff
@@ -74,4 +77,21 @@ default long getRemainingItemCount() {
     default long getStartItemCount() {
         return 0;
     }
+
+    /**
+     * @param craftCompleteListener a callback that is called when task is complete
+     */
+    void addOnCompleteListener(CraftCompleteListener craftCompleteListener);
+
+    /**
+     * @param onCancelListener a callback that is called when task is canceled
+     */
+    void addCancelListener(CraftCancelListener onCancelListener);
+
+    /**
+     * called when craft executes, passing number of tasks executed to Listener
+     * 
+     * @param onCraftingStatusUpdate
+     */
+    void addCraftingStatusListener(CraftingStatusListener onCraftingStatusUpdate);
```

Comment:
Please make sure that names of all methods and arguments are consistent. I'll be happy with any naming convention (either `addOn[Action]Listener(Craft[Action]Listener on[Action]Listener)` or `add[Action]Listener(Craft[Action]Listener craft[Action]Listener)`) but not mixed.
Also add default empty implementation for backward compatibility.